enhancements/installer/bootstrapping-clusters-with-capi-providers.md

```diff
@@ -0,0 +1,649 @@
+    scheme: internet-facing
+  network:
+    additionalControlPlaneIngressRules:
+    - description: MCS traffic from cluster network
+      fromPort: 22623
+      protocol: tcp
+      sourceSecurityGroupRoles:
+      - node
+      - controlplane
+      toPort: 22623
+    - description: controller-manager
+      fromPort: 10257
+      protocol: tcp
+      sourceSecurityGroupRoles:
+      - controlplane
+      - node
+      toPort: 10257
+    - description: kube-scheduler
+      fromPort: 10259
+      protocol: tcp
+      sourceSecurityGroupRoles:
+      - controlplane
+      - node
+      toPort: 10259
+    cni:
+      cniIngressRules:
+      - description: ICMP
+        fromPort: -1
+        protocol: icmp
+        toPort: -1
+       # several rules omitted for brevity
+      - description: Service node ports (UDP)
+        fromPort: 30000
+        protocol: udp
+        toPort: 32767
+    vpc:
+      availabilityZoneSelection: Ordered
+      availabilityZoneUsageLimit: 3
+      cidrBlock: 10.0.0.0/16
+  region: us-east-2
+  s3Bucket:
+    name: openshift-bootstrap-data-capi-example-kbdpg
+    presignedURLDuration: 1h0m0s
+status:
+  networkStatus:
+    apiServerElb:
+      attributes: {}
+  ready: false
+```
+
+This manifest describes the network, load balancers, security group rules, etc. that will be used or provisioned for the target cluster.
+Control plane machines are provisioned by [AWSMachines][AWSMachines] & [CAPI Machines][CAPIMachines] manifests.
+
+##### Additional Infrastructure
+
+The Installer generates any additional infrastructure that is needed but not handled by the CAPI provider, either because it is out of scope (e.g. IAM, DNS), or not adopted upstream (e.g. split-horizon load balancers).
+The Installer codebase provides hooks into the provisioning lifecycle that can be used to provision resources using direct SDK calls or other tooling.
+
+In most cases, teams are encouraged in discussing and building issues in the respective upstream repositories first, e.g. `cluster-api` or `cluster-api-provider-aws` and only
+use the following hook as a fallback path.
+
+Provisioning of resources with CAPI is divided into two stages: _infrastructure_ (all non-machine related manifests) and _control plane_,
+with hooks before, in between, and after:
+
+* PreProvision Hook - handle prequisites: e.g. IAM role creation, OS Disk upload
+* Provision Infrastructure - create `cluster` manifest on local control plane
+* InfraReady Hook - handle requirements in between cluster infrastructure and machine provisioning
+* Ignition Hook - similar to InfraReady but specifically for generating (bootstrap) ignition
+* Provision Machines - create bootstrap ignition and control-plane machines
+* PostProvision Hook - post-provision requirements, such as DNS record creation
+
+##### Bootstrap Resources
+
+OpenShift will continue to use the same bootstrapping process, as described in the Installer
+[Cluster Installation Process][install-overview].
```

Comment:
Note that for baremetal the bootstrap Machine is not the same type as cluster Machines (libvirt vs metal3). Does it pose a problem?

Reply:
Nothing comes to mind where this would be a problem in this design. I'm not sure of the implementation details, but the libvirt VM could be setup by the installer in one of the hooks.

Reply:
Yeah, that's what we'll probably settle on based on our discussions. I don't expect the baremetal platform to use CAPI at all for the time being.